### PR TITLE
Search rules

### PR DESCRIPTION
**Pull request by kasimebrahim**
as per https://github.com/singnet/ure/issues/16

### CODE REVIEW

**Comment by ngeiswei** [comment]
Thanks @kasimebrahim !

It's not clear to me yet if I should merge here or on singnet, we're technically supposed to work on singnet now, but they are some issues I'd like to clear up regarding circleci config first, so I'm gonna let it open in the meantime.

**Comment by linas** [comment]
why are we "technically supposed to work on signet now" and who is pushing this agenda? Is this still Cassio?

I despise the fragmentation and tumult that singnet caused, I think it was a truly evil strategy and I want to completely smash it into smithereens.

**Comment by ngeiswei** [comment]
@linas, I can see your viewpoint, and I also despise fragmentation, but there are other things to take into account like the fact you have more or less total control over opencog/atomspace, which I don't personally mind, but that doesn't give enough room for new folks like the StP team to experiment with their wild ideas. This among other things which are a bit over my head, but that's the main bit I believe.

Anyway, the technical issue mentioned is just to be able to hassle-free merge from singnet to opencog and vice versa. That's not gonna be possible for the atomspace repo (due to you not accepting changes like PtrValue as they are), but my intention is to keep all other/my repos on the opencog side up to date. So basically for the time being, as far as "my" repos are concerned opencog and singnet are gonna mirror each other. I just had to fix a few technical bits to make this happen (having to do with README and circleci pointing to the right circleci and docker urls). This is now fixed and I can take care of doing the various merging within the coming days.

File: opencog/scm/opencog/ure/ure-utils.scm [hunk 1]
```diff
@@ -537,6 +539,26 @@
     (cog-set-atomspace! current-as)
     clean-weighted-rules))
 
+(define*-public (ure-search-rules  #:optional (as (cog-atomspace)))
+"
+  Retrieve all potential rules in the given or default atomspace.
+  https://github.com/singnet/ure/issues/16
```
Comment (ngeiswei) [inline opencog/scm/opencog/ure/ure-utils.scm:545]: Ideally, you'd add a usage, cause doing `(help <function>)` only shows the comment, not the signature.

**Comment by ngeiswei** [comment]
I'm merging @kasimebrahim, if you want to fix the docstring you may do it in another PR. Thanks.

**Comment by linas** [comment]
I'm disappointed to read what you wrote.  The StP team made a fundamental technical error by choosing to place code into the atomspace, instead of just working in a brand new repo. For example, the `attentionbank`, and the `spacetime` server fly to mind - both define new atom and value types, both define new ways of interacting with the atomspace to store new kinds of information.  They could have done that .. they could still do that ... and both @vsbogd and @noskill understand that this is *still* a possibility. -- its not even that hard to do! A few days of work, maybe, to split out their work into it's own repo, and then we could be back to just having a single AtomSpace repo.

Why should `PtrValue` be split out into it's own repo? For all the usual reasons for modular design:
* Reduced software maintenance effort for both repos. The original authors of attentionbank and spacetime can do whatever they want, without anyone else getting involved. The maintainers of atomspace don't have to worry about getting broken.
* When subprojects die, and the maintainers leave, it is easier to mothball. For example, `dimensional embedding` is dead. Too bad: but the good news is that neither the atomspace, nor attentionbank, nor spacetime, nor the StP team, nor anyone else in the SingNet organzation have to worry about the death of `dimensional-embedding`. None of us have to continue to sweat bullets and curse the keyboard over it any longer -- a pain point is gone. Less pain is always better.
* Failed designs are more easily abandoned. I created an `atomsace-to-ipfs` shim, it got to version 0.1, and is now abandoned, because I realized the design was wrong. Great! The good news is no one has to remove thousands of lines of code from the SingNet AtomSpace repo or OpenCog AtomSpace repo! We don't even have to argue about whther or not to remove the code. Fewer arguments are better!

Above are three reasons, but ***the most important reason*** is the last: ***stability through project management***. The AtomSpace now has many users, and it needs to stay stable. It's not really a good place for "experimentation" any more. We have a very long historical pattern of having people show up randomly, be encouraged (because Ben likes them, and Ben encourages people, which is a good thing to do) but then they commit changes that break stuff, and then they leave. ... and so we get code that is both broken, and abandoned, and its wedged tightly into the core. As a software organization, we cannot afford to operate like that. There are too few of us, we are stretched too thin.

The answer: keep the atomspace thin and small -- the thinner and smaller, the better. The javascript people have discovered this as a way of life: Lots of itty-bitty single-function components that are complete, stable, work-as-documented. This is kind of harder in C++, but we can go part of the way there. There is no technical reason to create giant, bloated C++ monoliths that strain under their own weight.

Why the StP team made this technical mistake ... I presume its because they had no management.   Why this technical mistake continues to live on ... I don't know.  I presume they still have no management -- that Ben helicopters in from time to time, and they march to his orders. Since he doesn't see the point of fixing this flaw, it doesn't get fixed. No one spends any time fixing it.

> but my intention is to keep all other/my repos on the opencog side up to date. 

OK, this is contrary to the original plan (as presented to me by Cassio). That plan said that the singnet repos would be deployed on cloud servers, running customer applications. To do that, you want to have stability: run acceptance tests, so that some random borken commit to opencog would not accidentally take down all the servers and leave all the customers unable to connect.  So the idea was that opencog was where we could add new code, and make whatever changes, while the singnet repos would be stable, and would be updated only infrequently.

The above describes a rather traditional software dev model. We are not following it. It is imperfect, and has it's own problems, which is why devops was invented. But we are not doing devops, either! What are we doing? Why?

From where I sit, the singnet/opencog split seems to not actually solve any problems, nor achieve any particular utility ... it seems to be some bizarre anti-pattern that persists because "singnet" is a cool-sounding  name, and no other reason.

I wrote this very long reply because I am hoping that "someone in charge" reads it and gives it some thought.  I do not know who is in charge. Sure, the nominal leader is @bgoertzel but he is completely dettached from the development process, completely unaware and totally disinterested of what is going on.  Fine. So who is in charge? No one -- nobody here but us chickens... OK, well, if there is no one here managing operations, lets at least figure out what we need to do ourselves.

**Pull request merged**